compressors/fano_coder.py

```diff
@@ -0,0 +1,160 @@
+from core.prob_dist import ProbabilityDist
+
+
+class FanoTree(PrefixFreeTree):
+    def __init__(self, prob_dist):
+        super().__init__(prob_dist)
+        # sort the probability distribution
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        # initialize root node of Fano Tree
+        self.root_node = BinaryNode(id=None)
+
+        # build tree returns the root node
+        super().__init__(root_node=self.build_fano_tree(self.root_node, self.sorted_prob_dist))
+
+
+    @staticmethod
+    def criterion(dict_item):
+        """used for finding the symbol at which cumulative probability is nearest to 0.5
+        This symbol is used to split between left and right trees of Fano. We split right starting this symbol
+        to get most balanced tree.
+        """
+        key, value = dict_item
+        return abs(value - 0.5)
+
+    @staticmethod
+    def split_prob_dist_into_two(norm_sort_prob_dist):
+        """
+        Given a normalized and sorted probability distribution, split it into two sets of approximately equal
+        probabilities as described in Fano's algorithm.
+        """
+        # Get cumulative probability dict
+        cumulative_prob_dict = norm_sort_prob_dist.cumulative_prob_dict
+
+        # Split the tree into left and right nodes based on cumulative probability and recursively build
+        # FanoTree for these trees
+        recursive_prob_dicts = {
+            'left': {},
+            'right': {}
+        }
+
+        # all alphabets which are nearest to 0.5 cumulative probability in sorted list are put to left tree and rest
+        # to right, such that we have a most balanced probabilistic tree subsets.
+        min_diff_code = min(cumulative_prob_dict.items(), key=FanoTree.criterion)[0]
+        curr_dict = recursive_prob_dicts['left']
+        for s, cum_prob in cumulative_prob_dict.items():
+            if s == min_diff_code:
+                curr_dict = recursive_prob_dicts['right']
+            curr_dict.update({s: norm_sort_prob_dist.probability(s)})
+
+        return recursive_prob_dicts
+
+
+    @staticmethod
+    def build_fano_tree(root_node, norm_sort_prob_dist) -> BinaryNode:
+        """recursively build Fano Tree"""
+        # split the symbols into left and right half symbols such that the probability weights in each set is
+        # most balanced
+        recursive_prob_dicts = FanoTree.split_prob_dist_into_two(norm_sort_prob_dist)
+
+        # Call recursion
+        # if only 1 symbol in either left or right tree, just assign it as a child and we don't have to call recursion
+        for branch, curr_dict in recursive_prob_dicts.items():
+            norm_prob_dict = ProbabilityDist.normalize_prob_dict(curr_dict)
+
+            # if pointer is None, the object is not updated
+            # https://stackoverflow.com/questions/55777748/updating-none-value-does-not-reflect-in-the-object
```

Comment:
move this outside the loop, and add comment like 
```
# we initialize the right and left child here and later separately update/recurse on them in the loop below (the loop has just 2 steps: left and right. Initialization is important since if we leave these as None, then a pattern like var = root_node.left_child; var.id = new_id` won't work because `var` would be just `None` and not a pointer
```

Reply:
thanks for the detailed comment.